presto-main/src/main/java/com/facebook/presto/execution/buffer/SpoolingOutputBuffer.java

```diff
@@ -0,0 +1,775 @@
+
+        tempDataOperationContext = new TempDataOperationContext(Optional.empty(), taskId.getQueryId().toString(), Optional.empty(), new Identity("spooling-buffer", Optional.empty()));
+
+        state.compareAndSet(OPEN, NO_MORE_BUFFERS);
+    }
+
+    @Override
+    public OutputBufferInfo getInfo()
+    {
+        return new OutputBufferInfo(
+                "SPOOLING",
+                state.get(),
+                state.get().canAddBuffers(),
+                state.get().canAddPages(),
+                totalBufferedBytes.get(),
+                totalBufferedPages.get(),
+                totalRowsAdded.get(),
+                totalPagesAdded.get(),
+                ImmutableList.of());
+    }
+
+    @Override
+    public boolean isFinished()
+    {
+        return state.get() == BufferState.FINISHED;
+    }
+
+    @Override
+    public double getUtilization()
+    {
+        return totalInMemoryBytes.get() / (double) threshold.toBytes();
+    }
+
+    @Override
+    public boolean isOverutilized()
+    {
+        return totalInMemoryBytes.get() > threshold.toBytes();
+    }
+
+    @Override
+    public long getPeakMemoryUsage()
+    {
+        return peakMemoryUsage.get();
+    }
+
+    @Override
+    public ListenableFuture<?> isFull()
+    {
+        return immediateFuture(null);
+    }
+
+    @Override
+    public void addStateChangeListener(StateMachine.StateChangeListener<BufferState> stateChangeListener)
+    {
+        state.addStateChangeListener(stateChangeListener);
+    }
+
+    @Override
+    public void setOutputBuffers(OutputBuffers newOutputBuffers)
+    {
+        requireNonNull(newOutputBuffers, "newOutputBuffers is null");
+        checkArgument(outputBuffers.getType() == SPOOLING, "Invalid output buffers type");
+        checkArgument(outputBuffers.isNoMoreBufferIds(), "invalid noMoreBufferIds");
+
+        if (state.get().isTerminal() || outputBuffers.getVersion() >= newOutputBuffers.getVersion()) {
+            return;
+        }
+
+        outputBuffers.checkValidTransition(newOutputBuffers);
+    }
+
+    @Override
+    public void enqueue(Lifespan lifespan, List<SerializedPage> pages)
+    {
+        if (!state.get().canAddPages()) {
+            return;
+        }
+        PendingRead pendingRead;
+        synchronized (this) {
+            this.pages.addAll(pages);
+
+            long bytesAdded = getPagesSize(pages);
+            long pagesAdded = pages.size();
+
+            // update output buffer info
+            totalBufferedBytes.addAndGet(bytesAdded);
+            totalBufferedPages.addAndGet(pagesAdded);
+            totalPagesAdded.addAndGet(pagesAdded);
+            totalRowsAdded.addAndGet(getPagesRows(pages));
+
+            totalInMemoryBytes.addAndGet(bytesAdded);
+
+            totalPagesRemaining.addAndGet(pagesAdded);
+            peakMemoryUsage.accumulateAndGet(totalInMemoryBytes.get(), Math::max);
+
+            if (totalInMemoryBytes.get() > threshold.toBytes()) {
```

Comment:
Move this condition to flush(). It will avoid multiple flush()s if we get rid of the `synchronized (this)`. You can rename if to `flushIfNeeded()`